draft-irtf-cfrg-opaque.md

```diff
@@ -830,20 +763,31 @@ Steps:
 2. N = Unblind(input.data_blind, Z)
 3. y = Finalize(PwdU, N, "RFCXXXX")
 4. RwdU = Harden(y, params)
-5. credentials_key = HKDF(salt=0, IKM=RwdU, "CredentialsKey", Nk)
-6. exporter_key = HKDF(salt=0, IKM=RwdU, "ExporterKey", Nk)
-7. n = random(Nn) # generate a fresh nonce randomly
-8. Credentials C with (skU=skU, pkS=response.pkS)
-9. (ct, t) = Seal(credentials_key, n, "", C)
-10. EnvU = concat(n, ct, t)
-11. Create RegistrationUpload upload with envelope value EnvU
-11. Output (upload, exporter_key)
+5. Create Credentials C according to server policy
+6. pt = SerializeExtensions(C.secret_credentials)
+7. nonce = random(Nn)
+8. pseudorandom_pad = HKDF-Expand(RwdU, contact(nonce, "Pad"), len(pt))
+9. auth_key = HKDF-Expand(RwdU, contact(nonce, "AuthKey"), Nk)
+10. exporter_key = HKDF-Expand(RwdU, concat(nonce, "ExporterKey"), Nk)
+11. ct = xor(pt, pseudorandom_pad)
+12. auth_data = SerializeExtensions(C.cleartext_credentials)
+13. t = HMAC(auth_key, concat(nonce, ct, concat(auth_data, aad)))
+14. Create Envelope EnvU with (nonce, ct, auth_data, t)
+15. Create RegistrationUpload upload with envelope value (EnvU, pkU).
+16. Output (upload, exporter_key)
 ~~~
 
 [[RFC editor: please change "RFCXXXX" to the correct number before publication.]]
 
-The application may optionally include pk(skU), IdS, and IdU in the Credentials
-structure C at step 8.
+Applications MUST encrypt and authenticate skU, and MUST authenticate pkS.
+Secrecy of pkS is optional. If an application requires secrecy of pkS, this value
+SHOULD be included in the `Credentials.secret_credentials` list (step 5).
+Applications may optionally include pkU, IdU, or IdS in the
+`Credentials.secret_credentials` structure (step 5) if secrecy of these values
+is desired. Otherwise, if an application does not require secrecy for these values
+but does require authentication, they may be appended to `Credentials.cleartext_credentials`.
```

Comment:
We need to specify that *all* values in EnvU require authentication, only secrecy is optional (except for skU for which it is mandatory)

Reply:
I think that's already clear from context, no?